src/PSRule.Rules.Azure/rules/Azure.Resource.Rule.ps1

```diff
@@ -34,3 +34,18 @@ Rule 'Azure.ResourceGroup.Name' -Ref 'AZR-000168' -Type 'Microsoft.Resources/res
     # Can't end with period.
     $Assert.Match($PSRule, 'TargetName', '^[-\w\._\(\)]*[-\w_\(\)]$');
 }
+
+
+# Synopsis: Ensure all properties named `adminUsername` within a deployment are expressions (not literal strings)
+Rule 'Azure.Resource.Deployments.adminUsername' -Ref 'unsure' -Type -ne 'Microsoft.Resources/deployments' -Tag @{ release = 'GA'; ruleSet = '2022_06' } {  
+    $deployment = @($TargetObject);
+    $adminUsername = $Assert.HasFieldValue($deployment, 'properties.adminUsername')
```

Comment:
A deployment will have a `properties.template.resources` property. `resources` will be an array of resources. Within the `resources` array for each resource look into `properties` and then find any property called `adminUsername`. If the `adminUsername` property exists then it shouldn't be a literal value.

To do a nested search of properties, it would be easiest to do it through a function currently although as a side node it would be good to implement the decendant operator in PSRule to enable `properties..adminUsername` I might log this as a feature on the PSRule repo.

The nested search PowerShell code might look like this:

```powershell
function global:FindAdminUsername {
  param ([PSObject]$InputObject)
  process {
    foreach ($property in $InputObject.PSObject.properties) {
      if ($property.Name -eq 'adminUsername') {
        $property
      }
      else if ($property.Value -is [PSObject]) {
        FindAdminUsername -InputObject $property.Value;
      }
    }
  }
}
```